File: base/src/main/java/util/CharacterIterable.java
package util;

import androidx.annotation.NonNull;

import java.util.Iterator;

/**
 * Iterates over a string treating a surrogate pair and a grapheme cluster a single character.
 */
public final class CharacterIterable implements Iterable<String> {

  private final String string;

  public CharacterIterable(@NonNull String string) {
    this.string = string;
  }

  @Override
  public @NonNull Iterator<String> iterator() {
    return new CharacterIterator();
  }

  private class CharacterIterator implements Iterator<String> {
    private static final int UNINITIALIZED = -2;

    private final BreakIteratorCompat breakIterator;

    private int lastIndex = UNINITIALIZED;

    CharacterIterator() {
      this.breakIterator = new AndroidIcuBreakIterator(string);
    }

    @Override
    public boolean hasNext() {
      if (lastIndex == UNINITIALIZED) {
        lastIndex = breakIterator.first();
      }
      return !breakIterator.isDone(lastIndex);
    }

    @Override
    public String next() {
      int firstIndex = lastIndex;
      lastIndex = breakIterator.next();
      return string.substring(firstIndex, lastIndex);
    }
  }

  private interface BreakIteratorCompat {
    int first();

    int next();

    boolean isDone(int index);
  }

  /**
   * The BreakIteratorCompat implementation, delegating to `android.icu.text.BreakIterator`.
   * Handles grapheme clusters correctly. Sole implementation now (minSdk=26 always
   * satisfies its API-24 requirement); do not re-add an SDK_INT/@RequiresApi guard.
   */
  private static class AndroidIcuBreakIterator implements BreakIteratorCompat {
    private final android.icu.text.BreakIterator breakIterator = android.icu.text.BreakIterator.getCharacterInstance();

    public AndroidIcuBreakIterator(@NonNull String string) {
      breakIterator.setText(string);
    }

    @Override
    public int first() {
      return breakIterator.first();
    }

    @Override
    public int next() {
      return breakIterator.next();
    }

    @Override
    public boolean isDone(int index) {
      return index == android.icu.text.BreakIterator.DONE;
    }
  }
}

File: base/src/main/java/util/CharacterIterable.kt
package util

/**
 * Iterates over a string treating a surrogate pair and a grapheme cluster a single character.
 */
class CharacterIterable(private val string: String) : Iterable<String> {

    override fun iterator(): Iterator<String> = CharacterIterator()

    private inner class CharacterIterator : Iterator<String> {
        // Local grapheme breaker; distinct from the top-level util.BreakIteratorCompat.
        private val breakIterator: GraphemeBreaker = AndroidIcuGraphemeBreaker(string)

        private var lastIndex = UNINITIALIZED

        override fun hasNext(): Boolean {
            if (lastIndex == UNINITIALIZED) {
                lastIndex = breakIterator.first()
            }
            return !breakIterator.isDone(lastIndex)
        }

        override fun next(): String {
            val firstIndex = lastIndex
            lastIndex = breakIterator.next()
            return string.substring(firstIndex, lastIndex)
        }
    }

    private interface GraphemeBreaker {
        fun first(): Int

        fun next(): Int

        fun isDone(index: Int): Boolean
    }

    /**
     * The GraphemeBreaker implementation, delegating to `android.icu.text.BreakIterator`.
     * Handles grapheme clusters correctly. Sole implementation now (minSdk=26 always
     * satisfies its API-24 requirement); do not re-add an SDK_INT/@RequiresApi guard.
     */
    private class AndroidIcuGraphemeBreaker(string: String) : GraphemeBreaker {
        private val breakIterator = android.icu.text.BreakIterator.getCharacterInstance()

        init {
            breakIterator.setText(string)
        }

        override fun first(): Int = breakIterator.first()

        override fun next(): Int = breakIterator.next()

        override fun isDone(index: Int): Boolean = index == android.icu.text.BreakIterator.DONE
    }

    companion object {
        private const val UNINITIALIZED = -2
    }
}
